src/analysis/processing/qgsalgorithmrandomextract.cpp

```diff
@@ -110,22 +110,48 @@ QVariantMap QgsRandomExtractAlgorithm::processAlgorithm( const QVariantMap &para
   QVector< QgsFeatureId > fids( number );
   std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
 
-  double step = number > 0 ? 100.0 / number : 1;
-  int i = 0;
-
-  QgsFeature f;
+  QHash< QgsFeatureId, int > idsCount;
   for ( QgsFeatureId id : fids )
   {
     if ( feedback->isCanceled() )
     {
       break;
     }
 
-    source->getFeatures( QgsFeatureRequest( id ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks ).nextFeature( f );
+    if ( idsCount.contains( id ) )
+    {
+      idsCount.insert( id, idsCount.value( id ) + 1 );
+    }
+    else
+    {
+      idsCount.insert( id, 1 );
+    }
+  }
+
+  QgsFeatureIds ids = QSet< QgsFeatureId >::fromList( idsCount.keys() );
+  QgsFeatureIterator fit = source->getFeatures( QgsFeatureRequest().setFilterFids( ids ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+  sink->addFeatures( fit, QgsFeatureSink::FastInsert );
 
-    sink->addFeature( f, QgsFeatureSink::FastInsert );
-    feedback->setProgress( i * step );
-    i++;
+  // add duplicated features if any
+  if ( ids.size() != number )
+  {
+    QgsFeature f;
+    for ( auto it = idsCount.constBegin(); it != idsCount.constEnd(); ++it )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+
+      if ( it.value() > 1 )
+      {
+        source->getFeatures( QgsFeatureRequest( it.key() ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks ).nextFeature( f );
+        for ( int j = 1; j < it.value(); j++ )
+        {
+          sink->addFeature( f, QgsFeatureSink::FastInsert );
```

Comment:
I think all this could be simplified:
```
while ( fit.nextFeature( f ) )
{
   const int count = ids.value( f.id() );
   for ( int i = 0; i < count; ++i )
   {
       sink->addFeature( f, QgsFeatureSink::FastInsert );
   }
}
```
That should be nice and efficient, and avoids all multiple feature requests